File: services/rules/src/model/transaction.js
const { ARG_COUNT_ERROR } = require('../constants');

module.exports = function(
  id,
  ruleInstanceId,
  author,
  authorDeviceId,
  authorDeviceLatlng,
  authorRole,
  equilibriumTime,
  sumValue,
  transactionItems,
) {
  const requiredArgCount = 9;
  if (arguments.length != requiredArgCount) {
    // todo: test
    throw new Error(ARG_COUNT_ERROR);
  };
  return {
    id: id,
    rule_instance_id: ruleInstanceId,
    author: author,
    author_device_id: authorDeviceId,
    author_device_latlng: authorDeviceLatlng,
    author_role: authorRole,
    equilibrium_time: equilibriumTime,
    sum_value: sumValue,
    transaction_items: transactionItems,
  };
};

File: services/gopkg/types/transaction.go
package types

// Transaction is an envelope for TransactionItems
type Transaction struct {
	ID                 *ID                `json:"id,omitempty"`
	RuleInstanceID     *ID                `json:"rule_instance_id,omitempty"`
	Author             *string            `json:"author"`
	AuthorDeviceID     *string            `json:"author_device_id,omitempty"`
	AuthorDeviceLatlng *string            `json:"author_device_latlng,omitempty"` // pg point type
	AuthorRole         *string            `json:"author_role"`
	EquilibriumTime    *string            `json:"equilibrium_time"`
	SumValue           *string            `json:"sum_value"`
	TransactionItems   []*TransactionItem `json:"transaction_items"`
}
